### ISSUE TITLE
[FR] Log metrics with confidence intervals and include them in plots

### ISSUE THREAD

**Issue by mauvilsa** [mlflow/mlflow#3237]
## Willingness to contribute
The MLflow Community encourages new feature contributions. Would you or another member of your organization be willing to contribute an implementation of this feature (either as an MLflow Plugin or an enhancement to the MLflow code base)?

- [ ] Yes. I can contribute this feature independently.
- [X] Yes. I would be willing to contribute this feature with guidance from the MLflow community.
- [ ] No. I cannot contribute this feature at this time.

## Proposal Summary

Confidence intervals (most commonly 95%) can be considered a best practice when comparing metrics to determine which models are better than others, or that the estimation of the metrics are too uncertain to be able to conclude anything. Honestly it puzzles me why the support for confidence intervals is not more widespread in machine learning logging frameworks.

## Motivation
- What is the use case for this feature?
The results of any experiment can benefit by logging both point estimates of performance metrics along with confidence intervals to be able to interpret the uncertainty of the metrics.
- Why is this use case valuable to support for MLflow users in general?
Support confidence intervals allows a good practice any mlflow user.
- Why is this use case valuable to support for your project(s) or organization?
We routinely run experiments computing confidence intervals for the metrics. We also use mlflow, but it is quite frustrating that it does not support confidence intervals because we can't use mlflow to visualize the results that we get.
- Why is it currently difficult to achieve this use case? (please be as specific as possible about why related MLflow features and components are insufficient)
Confidence intervals are best used when the are included in plots as error bars which allows easy comparison between different models. Including the low and high values of the intervals as different metrics is not useful because it makes the plots confusing. The error bars are actually important to ease the interpretation of the results.

### What component(s), interfaces, languages, and integrations does this feature affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [X] `area/docs`: MLflow documentation pages
- [X] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interfaces
- [X] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Languages 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

## Details

Implementing this feature would require:
- Extend the log-metric functions in all APIs to support optional lower and upper limits of the confidence intervals.
- Extend the storage of metrics so that it can optionally include the intervals.
- Extend plotting functionalities in the ui so that if confidence intervals are available the user can choose to display them or not.

**Comment by harupy** [comment]
@mauvilsa Thanks for filing this.

> We also use mlflow, but it is quite frustrating that it does not support confidence intervals because we can't use mlflow to visualize the results that we get.

You can create a plot and log it as an artifact (we currently don't support comparing artifacts across runs though). Doesn't this approach work for you?

**Comment by mauvilsa** [comment]
@harupy indeed, we already use artifacts to store plots with confidence intervals. However, this is not good enough. The most important part of the feature request is to be able to compare across runs. Being able to compare artifacts across runs would also be an interesting feature to have. But using artifacts to compare metrics would not be very optimal. It is important to observe metrics on the same y-axis to easily judge if and how much intervals overlap for different runs. Doing this with artifacts would require you to create horizontally narrow plots with a single metric without any labels or tics. And then the comparison of artifacts would need to guarantee that all compared artifacts are perfectly aligned horizontally.

I know this is not a small feature which may not be implemented due to its size or lack of interest from more people. Though I do consider it important and wanted to through out the idea to see what happens.

**Comment by skylarbpayne** [comment]
I would really like this feature too, and would be willing to contribute it.
